src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -30,21 +32,432 @@
+            Assert.Equal(SourceCodeKind.Script, newSolution.GetDocument(documentId)!.SourceCodeKind);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithDocumentSyntaxRoot()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().DocumentIds.Single();
+            var root = CS.SyntaxFactory.ParseSyntaxTree("class NewClass {}").GetRoot();
+
+            var newSolution1 = solution.WithDocumentSyntaxRoot(documentId, root, PreservationMode.PreserveIdentity);
+            Assert.True(newSolution1.GetDocument(documentId)!.TryGetSyntaxRoot(out var actualRoot));
+            Assert.Equal(root.ToString(), actualRoot!.ToString());
+
+            // the actual root has a new parent SyntaxTree:
+            Assert.NotSame(root, actualRoot);
+
+            var newSolution2 = newSolution1.WithDocumentSyntaxRoot(documentId, actualRoot);
+            Assert.Same(newSolution1, newSolution2);
+
+            Assert.Throws<ArgumentOutOfRangeException>(() => solution.WithDocumentSyntaxRoot(documentId, root, (PreservationMode)(-1)));
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentSyntaxRoot(null!, root));
+            Assert.Throws<InvalidOperationException>(() => solution.WithDocumentSyntaxRoot(s_unrelatedDocumentId, root));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        [WorkItem(37125, "https://github.com/dotnet/roslyn/issues/41940")]
+        public void WithDocumentSyntaxRoot_AnalyzerConfigWithoutFilePath()
+        {
+            var projectId = ProjectId.CreateNewId();
+
+            var solution = CreateSolution()
+                .AddProject(projectId, "goo", "goo.dll", LanguageNames.CSharp)
+                .AddDocument(DocumentId.CreateNewId(projectId), "goo.cs", "public class Goo { }")
+                .AddAnalyzerConfigDocument(DocumentId.CreateNewId(projectId), "editorcfg", SourceText.From("config"));
+
+            var documentId = solution.Projects.Single().DocumentIds.Single();
+
+            var root = CS.SyntaxFactory.ParseSyntaxTree("class NewClass {}", path: "").GetRoot();
+            Assert.Throws<ArgumentException>(() => solution.WithDocumentSyntaxRoot(documentId, root));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithDocumentText_SourceText()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().DocumentIds.Single();
+            var text = SourceText.From("new text");
+
+            var newSolution1 = solution.WithDocumentText(documentId, text, PreservationMode.PreserveIdentity);
+            Assert.True(newSolution1.GetDocument(documentId)!.TryGetText(out var actualText));
+            Assert.Equal(text, actualText);
```

Comment:
```suggestion
            Assert.Same(text, actualText);
```

Since you're using PreserveIdentity?